File: backend/internal/core/domain/preparation.go
package domain

type BlockType string

const (
	BlockInstruction      BlockType = "instruction"
	BlockTeacherSpeech    BlockType = "teacher_speech"
	BlockExpectedAnswer   BlockType = "expected_answer"
	BlockTeacherRelaunch  BlockType = "teacher_relaunch"
	BlockAnticipatedError BlockType = "anticipated_error"
	BlockSupport          BlockType = "support"
	BlockExtension        BlockType = "extension"
)

var BlockTypes = []BlockType{
	BlockInstruction,
	BlockTeacherSpeech,
	BlockExpectedAnswer,
	BlockTeacherRelaunch,
	BlockAnticipatedError,
	BlockSupport,
	BlockExtension,
}

func (b BlockType) IsValid() bool {
	for _, known := range BlockTypes {
		if b == known {
			return true
		}
	}
	return false
}

type Block struct {
	Type BlockType
	Text string
}

type Phase struct {
	Name            string
	DurationMinutes int
	Organization    string
	Blocks          []Block
}

type Sheet struct {
	Title           string
	Subject         string
	Level           string
	DurationMinutes int
	Competencies    []string
	Objective       string
	Materials       []string
	Phases          []Phase
}

func (s Sheet) Valid() bool {
	if s.Title == "" || s.Subject == "" || s.Level == "" || s.Objective == "" {
		return false
	}
	if len(s.Phases) == 0 {
		return false
	}
	for _, phase := range s.Phases {
		if phase.Name == "" || phase.Organization == "" || len(phase.Blocks) == 0 {
			return false
		}
		for _, block := range phase.Blocks {
			if !block.Type.IsValid() || block.Text == "" {
				return false
			}
		}
	}
	return true
}

File: frontend/src/types/preparation.ts
export type BlockType =
  | 'instruction'
  | 'teacher_speech'
  | 'expected_answer'
  | 'teacher_relaunch'
  | 'anticipated_error'
  | 'support'
  | 'extension'

export type PreparationBlock = {
  type: BlockType
  text: string
}

export type PreparationPhase = {
  name: string
  durationMinutes: number
  organization: string
  blocks: PreparationBlock[]
}

export type PreparationSheet = {
  title: string
  subject: string
  level: string
  durationMinutes: number
  competencies: string[]
  objective: string
  materials: string[]
  phases: PreparationPhase[]
}

export type SavedSheet = {
  id: string
  createdAt: string
  sheet: PreparationSheet
}

export type SheetSummary = {
  id: string
  title: string
  subject: string
  level: string
  durationMinutes: number
  createdAt: string
}

export type ExtractedDocument = {
  name: string
  type: string
  text: string
}

export type GenerationMode = 'fast' | 'advanced'
export type SavedSheetSaveMode = 'replace' | 'copy'

export type CreateSheetPayload = {
  subject: string
  level: string
  durationMinutes: number
  resources: ExtractedDocument[]
  notes?: string
  availableMaterials?: string
  adaptToClass?: boolean
  period?: string
  generationMode?: GenerationMode
}

export type ImproveSheetPayload = {
  existingSheet: ExtractedDocument
  resources: ExtractedDocument[]
  notes?: string
  availableMaterials?: string
  adaptToClass?: boolean
  generationMode?: GenerationMode
}

export type ImproveSavedSheetPayload = {
  resources: ExtractedDocument[]
  notes?: string
  availableMaterials?: string
  adaptToClass?: boolean
  generationMode?: GenerationMode
  saveMode?: SavedSheetSaveMode
}

export type UpdateSheetPayload = PreparationSheet

export type NeedGroup = {
  name: string
  needs: string
  adaptations: string
}

export type StudentProfile = {
  name: string
  strengths: string
  difficulties: string
  needs: string
  adaptations: string
}

export type ClassProfile = {
  level: string
  studentCount: number
  overallLevel: string
  classroomContext: string
  defaultMaterials: string
  avoidMaterials: string
  preferredSupports: string[]
  defaultSessionDuration: number
  pedagogicalPreferences: string[]
  needGroups: NeedGroup[]
  students: StudentProfile[]
  updatedAt?: string
}

export type AuthUser = {
  id: string
  email: string
}

export type AuthResponse = {
  token: string
  user: AuthUser
}

export type UpdateProfilePayload = {
  email: string
  password?: string
}
